service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,101 +22,341 @@ package sysworkflow
+		logging.TagArchiveRequestWorkflowID:        request.WorkflowID,
+		logging.TagArchiveRequestRunID:             request.RunID,
+		logging.TagArchiveRequestEventStoreVersion: request.EventStoreVersion,
+		logging.TagArchiveRequestLastFirstEventID:  request.LastFirstEventID,
+	})
+	metricsClient := container.MetricsClient
+	domainCache := container.DomainCache
+	clusterMetadata := container.ClusterMetadata
+	domainCacheEntry, err := domainCache.GetDomainByID(request.DomainID)
+	if err != nil {
+		logger.WithError(err).Error("failed to get domain from domain cache")
+		metricsClient.IncCounter(metrics.ArchivalUploadActivityScope, metrics.SysWorkerGetDomainFailures)
+		return errArchivalUploadActivityGetDomain
+	}
+	if !clusterMetadata.IsArchivalEnabled() {
+		// for now if archival is disabled simply abort the activity
+		// a more in depth design meeting is needed to decide the correct way to handle backfilling/pausing archival
+		logger.Warn("archival is not enabled for cluster, skipping archival upload")
+		metricsClient.IncCounter(metrics.ArchivalUploadActivityScope, metrics.SysWorkerArchivalNotEnabledForCluster)
+		return nil
+	}
+	if domainCacheEntry.GetConfig().ArchivalStatus != shared.ArchivalStatusEnabled {
+		// for now if archival is disabled simply abort the activity
+		// a more in depth design meeting is needed to decide the correct way to handle backfilling/pausing archival
+		logger.Warn("archival is not enabled for domain, skipping archival upload")
+		metricsClient.IncCounter(metrics.ArchivalUploadActivityScope, metrics.SysWorkerArchivalNotEnabledForDomain)
+		return nil
+	}
+	historyBlobItr := NewHistoryBlobIterator(
+		logger,
+		metricsClient,
+		container.HistoryManager,
+		container.HistoryV2Manager,
+		request.DomainID,
+		request.WorkflowID,
+		request.RunID,
+		request.EventStoreVersion,
+		request.BranchToken,
+		request.LastFirstEventID,
+		container.Config,
+		domainCacheEntry.GetInfo().Name,
+		container.ClusterMetadata.GetCurrentClusterName(),
+		request.CloseFailoverVersion,
+	)
+
+	blobstoreClient := container.Blobstore
+	bucket := domainCacheEntry.GetConfig().ArchivalBucket
+	for historyBlobItr.HasNext() {
+		historyBlob, err := nextBlobRetryForever(historyBlobItr)
+		if err != nil {
+			logger.WithError(err).Error("failed to get next blob from iterator, stopping archival upload")
+			metricsClient.IncCounter(metrics.ArchivalUploadActivityScope, metrics.SysWorkerNextBlobNonRetryableFailures)
+			return errArchivalUploadActivityNextBlob
+		}
+		key, err := NewHistoryBlobKey(request.DomainID, request.WorkflowID, request.RunID, *historyBlob.Header.CurrentPageToken)
+		if err != nil {
+			logger.WithError(err).Error("failed to construct blob key, stopping archival upload")
+			metricsClient.IncCounter(metrics.ArchivalUploadActivityScope, metrics.SysWorkerKeyConstructionFailures)
+			return errArchivalUploadActivityConstructKey
+		}
+		if exists, err := blobExistsRetryForever(blobstoreClient, bucket, key); err != nil {
```

Comment:
If this retries forever, you want to make sure you heartbeat periodically to avoid an activity heartbeat timeout

Reply:
Same response as above. It seems like we are thinking about heartbeating differently. Let's chat in person about the right model here.